priority-messages/src/main/java/io/streamnative/pulsar/recipes/client/priority/PriorityConsumerInterceptor.java

```diff
@@ -0,0 +1,127 @@
+        String priorityString = message.getProperties().get(ConstsForPriorityMessages.MSG_PROP_PRIORITY);
+        // If no property in message, means the lowest priority.
+        if (StringUtils.isBlank(priorityString)) {
+            return Integer.MAX_VALUE;
+        }
+        if (!StringUtils.isNumeric(priorityString)) {
+            log.warn("The priority prop [{}] of message is not a number.", ConstsForPriorityMessages.MSG_PROP_PRIORITY);
+            return Integer.MAX_VALUE;
+        }
+        return Integer.valueOf(priorityString);
+    }
+
+    /**
+     * If the high priority partition has many messages, it suspends receiving messages from the low priority partition.
+     * Note: If some messages with low priority have already been received into the memory, they will be consumed as
+     * high level priority.
+     */
+    private synchronized void triggerPauseOrResume() {
+        if (consumerMapping.size() == 1){
+            return;
+        }
+
+        boolean doPause = false;
+        for (Map.Entry<Integer, Set<ConsumerImpl<T>>> entry : consumerMapping.entrySet()){
+            if (doPause){
+                entry.getValue().forEach(ConsumerImpl::pause);
+                continue;
+            }
+            entry.getValue().forEach(ConsumerImpl::resume);
+            long messageCountInMemory = entry.getValue().stream().map(ConsumerImpl::getTotalIncomingMessages)
+                    .reduce((s1, s2) -> s1 + s2).get();
+            if (messageCountInMemory > 50){
+                doPause = true;
+            }
+        }
+    }
+
+    @Override
+    public Message<T> beforeConsume(Consumer<T> consumer, Message<T> message) {
+        // This method is just used to collect the consumers created by the client for each partition.
+        ConsumerImpl<T> consumerImpl = (ConsumerImpl<T>) consumer;
+        if (consumerSet.contains(consumerImpl.getTopic())){
+            return message;
+        }
+        // Registry consumer.
+        int priority = calculatePriority(message);
+        consumerMapping.computeIfAbsent(priority, i -> Collections.synchronizedSet(new HashSet<>()));
+        consumerMapping.get(priority).add(consumerImpl);
+        consumerSet.add(consumerImpl.getTopic());
+        return message;
+    }
+
+    @Override
+    public void close() {
+        consumerMapping.clear();
+        consumerSet.clear();
+    }
+
+    @Override
+    public void onAcknowledge(Consumer<T> consumer, MessageId messageId, Throwable exception) {
+        triggerPauseOrResume();
+    }
+
+    @Override
+    public void onAcknowledgeCumulative(Consumer<T> consumer, MessageId messageId, Throwable exception) {
+        triggerPauseOrResume();
+    }
+
+    @Override
+    public void onNegativeAcksSend(Consumer<T> consumer, Set<MessageId> messageIds) {
+        triggerPauseOrResume();
+    }
+
+    @Override
+    public void onAckTimeoutSend(Consumer<T> consumer, Set<MessageId> messageIds) {
+        triggerPauseOrResume();
+    }
```

Comment:
For this solution, if users won't call ack or nack for many messages. The priority delivery will not be applied, right?

Reply:
> For this solution, if users won't call ack or nack for many messages. The priority delivery will not be applied, right?

Yes. We can explain this scenario in other words: the messages received in the first round will not be prioritized.